_posts/2020-03-15-method-s-arguments-pt-1.markdown

```diff
@@ -0,0 +1,372 @@
+- Methods are always executed in the context of an object, so we can call a method with no arguments but we always have access to the method's state and we can modify it.
+
+In Ruby, where everything is an object, we can never have real functions but we can have methods that won't use or modify the object's state if we want to have something similar.
+
+# Arguments Vs. Parameters
+
+Another thing that's worth to differenciate is the concept of arguments and parameters. When we define a method, the parameters are part of the definition, the parameters are the variables that will contain the values that we use when calling that method. Arguments are the actual values that we use when we call a method. So, in this code:
+
+```ruby
+def my_method(foo)
+  # foo is a parameter when defining the method
+end
+
+my_method("bar")
+# "baz" is the argument we use when calling the method
+```
+
+Most of the times I'll use the word "arguments" for simplicity, since a method defines the parameters and accepts arguments.
+
+# No arguments
+
+We can have methods that have no input, like this:
+
+```ruby
+def pi # parenthesis are optional and the keyword "return" is optional too
+  3.14
+end
+```
+
+But given that this is a method, it has access to the object's state to do different things
+
+```ruby
+class Circle
+  def initialize(radius)
+    @radius = radius
+  end
+
+  def pi
+    3.14
+  end
+
+  def area
+    pi * @radius ** 2
+  end
+end
+
+my_circle = Circle.new(3)
+puts my_circle.area
+# => 28.26
+
+other_circle = Circle.new(4)
+puts other_circle.area
+# => 50.24
+```
+
+# Positional arguments
+
+This type of arguments get that name because the order you use matters. You can have 0 or more positional arguments, and you can have required and optional arguments.
+
+# Required positional arguments
+
+This positional arguments are required when calling a method, it's not that it's required to used them when defining the method (you can have no arguments, only optionals, etc). If the method defines required parameters you have to provide an argument for each position.
+
+```ruby
+def foo(arg1)
+  puts "arg1 is: #{arg1.inspect}"
+end
+
+foo(1)
+# => arg1 is: 1
+
+foo
+# ArgumentError (wrong number of arguments (given 0, expected 1))
+```
+
+We can have as many as we want:
+
+```ruby
+def foo(arg1, arg2, arg3, arg4)
+# and more and more...
+```
+
+> Note that methods with too many arguments are a sign of a bad design since the method is probably doing too many things!
+
+# Optional positional arguments
+
+Some times we may want to allow the user to provide more input but not require them to do so. For that, we can use optional arguments for which we define a default value in case the user didn't specify one. This is particularly usefull when we want to have some default behavior for our method but give the user the option to modify that, if we just hardcode the default value inside the method we wouldn't have this flexibility.
```

Comment:
```suggestion
Some times we may want to allow the user to provide more input but not require them to do so. For that, we can use optional arguments for which we define a default value in case the user doesn't specify one. This is particularly useful when we want to have some default behavior for our method but give the user the option to modify that. If we just hardcoded the default value inside the method we wouldn't have this flexibility.
```